medcat/utils/relation_extraction/rel_dataset.py

```diff
@@ -0,0 +1,687 @@
+                else:
+                    left_context_start_char_pos = tokenizer_data[
+                        "offset_mapping"][ent1_left_ent_context_token_pos_end][0]
+
+                ent2_right_ent_context_token_pos_end = ent2_token_start_pos + \
+                    self.config.general.cntx_right
+
+                # get end of 2nd ent token (if using tags)
+                if self.config.general.annotation_schema_tag_ids:
+                    far_pos = -1
+                    for tkn_id in self.config.general.annotation_schema_tag_ids:
+                        pos = [i for i in range(
+                            0, doc_length) if tokenizer_data["input_ids"][i] == tkn_id][0]
+                        far_pos = pos if far_pos < pos else far_pos
+                    ent2_right_ent_context_token_pos_end = far_pos
+
+                if ent2_right_ent_context_token_pos_end >= doc_length - 1:
+                    ent2_right_ent_context_token_pos_end = doc_length - 2
+                else:
+                    right_context_start_end_pos = tokenizer_data[
+                        "offset_mapping"][ent2_right_ent_context_token_pos_end][1]
+
+                ent1_token = tokenizer_data["tokens"][ent1_token_start_pos]
+                ent2_token = tokenizer_data["tokens"][ent2_token_start_pos]
+
+                window_tokenizer_data = self.tokenizer(
+                    doc_text[left_context_start_char_pos:right_context_start_end_pos])
+
+                # update token loc to match new selection
+                if self.config.general.annotation_schema_tag_ids:
+                    ent1_token_start_pos = \
+                        window_tokenizer_data["input_ids"].index(
+                            self.config.general.annotation_schema_tag_ids[0])
+                    ent2_token_start_pos = \
+                        window_tokenizer_data["input_ids"].index(
+                            self.config.general.annotation_schema_tag_ids[2])
+                else:
+                    ent2_token_start_pos = ent2_token_start_pos - ent1_token_start_pos
+                    ent1_token_start_pos = self.config.general.cntx_left if ent1_token_start_pos - \
+                        self.config.general.cntx_left > 0 else ent1_token_start_pos
+                    ent2_token_start_pos += ent1_token_start_pos
+
+                ent1_ent2_new_start = (
+                    ent1_token_start_pos, ent2_token_start_pos)
+
+                en1_start, en1_end = window_tokenizer_data["offset_mapping"][ent1_token_start_pos]
+                en2_start, en2_end = window_tokenizer_data["offset_mapping"][ent2_token_start_pos]
+
+                relation_instances.append([window_tokenizer_data["input_ids"], ent1_ent2_new_start, ent1_token, ent2_token, "UNK", self.config.model.padding_idx,
+                                           None, None, None, None, None, None, doc_id, "",
+                                           en1_start, en1_end, en2_start, en2_end])
+
```

Comment:
these lines appeared repeated across the plain text, `Spacy.Doc` impl lines 363-393, and the `create_relations_from_export` lines 588-614 ?

Reply:
Happy for you to mark as cleanup required - and get this merged in so we can start testing / using the impl. If we address later can you create a ticket and comment in here with details.

Reply:
I will rewrite the create_from* methods because they do contain repetitive bits + will be harder to maintain, just wanted to get this working and ready.